corehq/apps/userreports/reports/view.py

```diff
@@ -437,7 +442,47 @@ def excel_response(self):
 
     @property
     @memoized
+    def export_too_large(self):
+        data = self.data_source
+        data.set_filter_values(self.filter_values)
+        total_rows = data.get_total_records()
+        return total_rows > UCR_EXPORT_TO_EXCEL_ROW_LIMIT
+
+    @property
+    @memoized
+    def export_size_check_response(self):
+        try:
+            too_large = self.export_too_large
+        except UserReportsError as e:
+            if settings.DEBUG:
+                raise
+            return self.render_json_response({
+                'export_allowed': False,
+                'message': e.message,
+            })
+
+        if too_large:
+            return self.render_json_response({
+                'export_allowed': False,
+                'message': _(
+                    "Report export is limited to {number} rows. "
+                    "Please filter the data in your report to "
+                    "{number} or fewer rows before exporting"
+                ).format(number=UCR_EXPORT_TO_EXCEL_ROW_LIMIT),
+            })
+        return self.render_json_response({
+            "export_allowed": True,
+        })
+
+    @property
+    @memoized
     def export_response(self):
+        if self.export_too_large:
+            # Frontend should check size with export_size_check_response()
+            # Before hitting this endpoint, but we check the size again here
+            # in case the user modifies the url manually.
+            raise Http404
```

Comment:
I think you're right to check, but I would raise `Http400`, because the page/resource is found, but the request is bad.